packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,378 @@
+                    pendingRegistryForElement.delete(element);
+                    internalUpgrade(element, registeredDefinition, definition);
+                }
+            }
+        }
+        // Flush whenDefined callbacks
+        const resolver = _definedResolvers.get(tagName);
+        if (resolver) {
+            resolver(constructor);
+        }
+    };
+
+    CustomElementRegistry.prototype.get = function get(
+        this: CustomElementRegistry,
+        tagName: string
+    ): any {
+        return nativeGet.apply(this, arguments) && _definitionsByTag.get(tagName)?.UserCtor;
+    };
+
+    CustomElementRegistry.prototype.whenDefined = function whenDefined(
+        this: CustomElementRegistry,
+        tagName: string
+    ): Promise<CustomElementConstructor> {
+        return nativeWhenDefined.apply(this, arguments).then(() => {
+            let promise = _definedPromises.get(tagName);
+            if (!promise) {
+                const definition = _definitionsByTag.get(tagName);
+                if (definition) {
+                    return Promise.resolve(definition.UserCtor);
+                }
+                let resolve: (constructor: CustomElementConstructor) => void;
+                promise = new Promise((r) => (resolve = r));
+                _definedPromises.set(tagName, promise);
+                _definedResolvers.set(tagName, resolve);
+            }
+            return promise;
+        });
+    };
+
+    // User extends this HTMLElement, which returns the CE being upgraded
+    let upgradingInstance: HTMLElement | undefined;
+
+    window.HTMLElement = function HTMLElement(this: HTMLElement) {
+        // Upgrading case: the pivoting class constructor was run by the browser's
+        // native custom elements and we're in the process of running the
+        // "constructor-call trick" on the natively constructed instance, so just
+        // return that here
+        const instance = upgradingInstance;
+        if (instance) {
+            upgradingInstance = undefined;
+            return instance;
+        }
+        // Construction case: we need to construct the pivoting instance and return it
+        // This is possible when the user register it via global registry and instantiate
+        // it via `new Ctor()`.
+        const { constructor } = this;
+        const definition = _definitionsByClass.get(constructor as CustomElementConstructor);
+        if (!definition || !definition.PivotCtor) {
+            throw new TypeError(
+                'Illegal constructor (custom element class must be registered with global customElements registry to be newable)'
+            );
+        }
+        // This constructor is ONLY invoked when it is the user instantiating
+        // an element via new Ctor while Ctor is a registered global constructor.
+        const { PivotCtor, UserCtor } = definition;
+        return new PivotCtor(UserCtor);
+    };
+    window.HTMLElement.prototype = NativeHTMLElement.prototype;
```

Comment:
AIUI, the `@lwc/engine-dom` package doesn't modify any globals (whereas `@lwc/synthetic-shadow` modifies lots). Is there a risk to having LWC get into the business of modifying the global `window.HTMLElement` here?

(Same question for `window.customElements` as well.)

Reply:
@caridy I was thinking about this some more... It seems to me that the global patches should ideally go in a separate package. We do the same thing for `@lwc/synthetic-shadow`. A few benefits:

- `@lwc/engine-dom` remains pure, doesn't override globals
- OSS consumers who don't need this behavior (i.e who are comfortable with WCs living in the same namespace) don't have to get it
- if/when we can get the behavior we need from the web platform (e.g. scoped custom elements registry), we can remove the polyfill

What do you think?

Reply:
I'm fine either way... but if you move it to another package, you need to have ways to interact with the internals of that other package, so, it is not really a polyfill, it has to have some backdoor for lwc to register internal stuff.

Reply:
also, the DOM package is about the DOM, and this is about the DOM as well, I don't see why not patching is really a problem for this case since the change is not observable for others.

Reply:
I just realized we are already patching globals in `@lwc/engine-dom`, and it seems to be for the same reason as this PR (lack of a usable scoped custom element registry):

https://github.com/salesforce/lwc/blob/6ebaaf1a98a20a59a1b6a62cb85d2399b43d0c26/packages/%40lwc/engine-dom/src/apis/create-element.ts#L39-L42

I guess ideally we should have all these global mutations in the same place? `@lwc/scoped-custom-element-registry-polyfill` or something like that?

Reply:
yeah.